server/plugin/plugin.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package plugin
+
+import (
+	"fmt"
+	gohttp "net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"text/template"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+
+	"github.com/mattermost/mattermost-plugin-msoffice/server/command"
+	"github.com/mattermost/mattermost-plugin-msoffice/server/config"
+	"github.com/mattermost/mattermost-plugin-msoffice/server/http"
+	"github.com/mattermost/mattermost-plugin-msoffice/server/kvstore"
+	"github.com/mattermost/mattermost-plugin-msoffice/server/user"
+	"github.com/mattermost/mattermost-plugin-msoffice/server/utils"
+)
+
+type Plugin struct {
+	plugin.MattermostPlugin
+	configLock  *sync.RWMutex
+	config      *config.Config
+	httpHandler *http.Handler
+
+	KVStore          kvstore.KVStore
+	UserStore        user.Store
+	OAuth2StateStore user.OAuth2StateStore
+
+	Templates map[string]*template.Template
+}
+
+func NewWithConfig(conf *config.Config) *Plugin {
+	return &Plugin{
+		configLock: &sync.RWMutex{},
+		config:     conf,
+	}
+}
+
+func (p *Plugin) OnActivate() error {
+	// if p.Templates == nil {
+	// 	templatesPath := filepath.Join(*(p.API.GetConfig().PluginSettings.Directory),
+	// 		p.config.PluginId, "server", "dist", "templates")
+	// 	templates, err := loadTemplates(templatesPath)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// 	p.Templates = templates
+	// }
+
+	kv := kvstore.NewPluginStore(p.API)
+	p.KVStore = kv
+	p.UserStore = user.NewStore(kv)
+	p.OAuth2StateStore = user.NewOAuth2StateStore(p.API)
+
+	command.Register(p.API.RegisterCommand)
+	p.httpHandler = p.newHTTPHandler(&(*p.config))
+
+	p.API.LogInfo(p.config.PluginId + " activated")
+	return nil
+}
+
+// OnConfigurationChange is invoked when configuration changes may have been made.
+func (p *Plugin) OnConfigurationChange() error {
+	conf := p.getConfig()
+	oldStored := conf.StoredConfig
+	newStored := config.StoredConfig{}
+	err := p.API.LoadPluginConfiguration(&newStored)
+	if err != nil {
+		return errors.WithMessage(err, "failed to load plugin configuration")
+	}
+
+	if newStored.OAuth2Authority == "" ||
+		newStored.OAuth2ClientId == "" ||
+		newStored.OAuth2ClientSecret == "" {
+		return errors.WithMessage(err, "failed to configure: OAuth2 credentials to be set in the config")
+	}
+
+	botUserId := conf.BotUserId
+	if newStored.BotUserName != oldStored.BotUserName {
+		user, appErr := p.API.GetUserByUsername(newStored.BotUserName)
+		if appErr != nil {
+			return errors.WithMessage(appErr, fmt.Sprintf("unable to load user %s", newStored.BotUserName))
+		}
+		botUserId = user.Id
+	}
+
+	mattermostSiteURL := *p.API.GetConfig().ServiceSettings.SiteURL
```

Comment:
I'm concerned about this error'ing out if the `Site URL` is not set, with no appropriate error shown.